internal/pkg/configmanager/app.go

```diff
@@ -1,84 +1,118 @@
+)
+
 type (
-	App struct {
-		Name         string             `yaml:"name" mapstructure:"name"`
-		Spec         RawOptions         `yaml:"spec" mapstructure:"spec"`
-		Repo         *Repo              `yaml:"repo" mapstructure:"repo"`
-		RepoTemplate *RepoTemplate      `yaml:"repoTemplate" mapstructure:"repoTemplate"`
-		CIPipelines  []PipelineTemplate `yaml:"ci" mapstructure:"ci"`
-		CDPipelines  []PipelineTemplate `yaml:"cd" mapstructure:"cd"`
+	appRaw struct {
+		Name         string         `yaml:"name" mapstructure:"name"`
+		Spec         map[string]any `yaml:"spec" mapstructure:"spec"`
+		Repo         *scm.SCMInfo   `yaml:"repo" mapstructure:"repo"`
+		RepoTemplate *scm.SCMInfo   `yaml:"repoTemplate" mapstructure:"repoTemplate"`
+		CIRawConfigs []pipelineRaw  `yaml:"ci" mapstructure:"ci"`
+		CDRawConfigs []pipelineRaw  `yaml:"cd" mapstructure:"cd"`
 	}
-	Apps []App
+)
 
-	// AppInConfig is the raw structured data in config file.
-	// The main difference between App and AppInConfig is "CI" and "CD" field.
-	// The "CIRawConfigs" is the raw data of "CI" field defined in config file,
-	// which will be rendered to "CIPipelines" field in App with "PipelineTemplates".
-	// The "CDRawConfigs" is similar to "CIRawConfigs".
-	AppInConfig struct {
-		Name         string        `yaml:"name" mapstructure:"name"`
-		Spec         RawOptions    `yaml:"spec" mapstructure:"spec"`
-		Repo         *Repo         `yaml:"repo" mapstructure:"repo"`
-		RepoTemplate *RepoTemplate `yaml:"repoTemplate" mapstructure:"repoTemplate"`
-		CIRawConfigs []CICD        `yaml:"ci" mapstructure:"ci"`
-		CDRawConfigs []CICD        `yaml:"cd" mapstructure:"cd"`
+// getAppTools return app tools
+func getAppTools(appStr string, globalVars map[string]any, templateMap map[string]string) (Tools, error) {
+	//1. render appStr with globalVars
+	appRenderStr, err := renderConfigWithVariables(appStr, globalVars)
+	if err != nil {
+		log.Debugf("configmanager/app %s render globalVars %+v failed", appRenderStr, globalVars)
+		return nil, fmt.Errorf("app render globalVars failed: %w", err)
 	}
-
-	CICD struct {
-		Type         string     `yaml:"type" mapstructure:"type"`
-		TemplateName string     `yaml:"templateName" mapstructure:"templateName"`
-		Options      RawOptions `yaml:"options" mapstructure:"options"`
-		Vars         RawOptions `yaml:"vars" mapstructure:"vars"`
+	// 2. unmarshal appRaw config for render pipelineTemplate
+	var rawData appRaw
+	if err := yaml.Unmarshal([]byte(appRenderStr), &rawData); err != nil {
+		return nil, fmt.Errorf("app parse yaml failed: %w", err)
 	}
-)
+	rawData.setDefault()
+	appVars := mapz.MergeMaps(globalVars, rawData.Spec)
+	// 3. generate app repo and tempalte repo from scmInfo
+	repoScaffoldingTool, err := rawData.getRepoTemplateTool(appVars)
+	if err != nil {
+		return nil, fmt.Errorf("app[%s] get repo failed: %w", rawData.Name, err)
+	}
+	// 4. get ci/cd pipelineTemplates
+	tools, err := rawData.getPipelineTools(templateMap, appVars)
```

Comment:
Same as above; generateCICDToolsFromAppConfig? In this way it's consistent with the comment at line 59.